### PR TITLE
Metrics and Events don't require attributes

### PR DESCRIPTION
**Pull request by jerbly**
Fix for issue #405 
- Updated the EBNF and JSON schema to define the `extends` or `attributes` requirement mandatory for all group types except `metric` and `event`.
- Added a group validity check as a warning.

### CODE REVIEW

**Review by lquerel — APPROVED** [review]
LGTM. Thanks

**Review by lmolkova — APPROVED** [review]
LGTM and thank you!

**Comment by lquerel** [comment]
@jerbly This PR is still in draft mode. Is that intentional? If not, please click “Ready for Review,” and I’ll proceed with merging it. Thanks!

**Comment by jerbly** [comment]
@lquerel it's currently failing unrelated clippy checks. One is a simple fix. The other is documentation and formatting in the generated code in weaver-codegen - I'm not sure what to do with it. Perhaps change clippy settings in that crate?

**Comment by lquerel** [comment]
Regarding Clippy, I suggest implementing the changes proposed in Clippy’s report regarding the possibility of eliding certain lifetimes. Currently, our build pipeline is configured to always use the latest version of the Rust toolchain, which has the advantage of benefiting from the latest changes and improvements but also the drawback of introducing new Clippy errors unrelated to the PR. We could pin the toolchain version to avoid this, but at the same time, the current approach ensures we don’t forget to update it.

**Comment by jerbly** [comment]
The lifetimes one is easy. The doc problems in the generated code was where I had my doubts it was worth the effort. I'll work on it.

**Comment by jerbly** [comment]
@lquerel - I've fixed those issues so this is ready to merge now. :)

File: crates/weaver_codegen_test/templates/registry/rust/attributes/mod.rs.j2 [hunk 1]
```diff
@@ -10,11 +9,12 @@
...
 use opentelemetry::{Key, KeyValue, StringValue};
 
+/// A constant indicating whether the attributes are enabled.
 pub const ATTRIBUTES_PARAM: bool = {{ params.attributes }};
 
 {% for attributes_root_namespace in ctx %}
-/// Attributes for the `{{ attributes_root_namespace.root_namespace }}` namespace.
-pub mod {{ attributes_root_namespace.root_namespace | snake_case }};
+//! Attributes for the `{{ attributes_root_namespace.root_namespace }}` namespace.
+//! pub mod {{ attributes_root_namespace.root_namespace | snake_case }};
```
Comment (lquerel) [inline crates/weaver_codegen_test/templates/registry/rust/attributes/mod.rs.j2:17]: Sorry, I didn’t understand these changes. Why comment out this part of the code generation?

File: crates/weaver_codegen_test/templates/registry/rust/metrics/mod.rs.j2 [hunk 2]
```diff
@@ -8,11 +7,12 @@
...
 //! DO NOT EDIT, THIS FILE HAS BEEN GENERATED BY WEAVER
 
+/// A constant indicating whether the metrics are enabled.
 pub const METRICS_PARAM: bool = {{ params.metrics }};
 
 {% for attributes_root_namespace in ctx %}
-/// Metrics for the `{{ attributes_root_namespace.root_namespace }}` namespace.
-pub mod {{ attributes_root_namespace.root_namespace | snake_case }};
+//! Metrics for the `{{ attributes_root_namespace.root_namespace }}` namespace.
+//! pub mod {{ attributes_root_namespace.root_namespace | snake_case }};
```
Comment (lquerel) [inline crates/weaver_codegen_test/templates/registry/rust/metrics/mod.rs.j2:15]: Same question.

**Review by lquerel — COMMENTED** [review]
I missed something but I didn’t understand the changes to the code generation section. What was the issue exactly? Why comment out that part of the code generation?

File: crates/weaver_codegen_test/templates/registry/rust/attributes/mod.rs.j2 (hunk 1 above)
Comment (jerbly) [reply crates/weaver_codegen_test/templates/registry/rust/attributes/mod.rs.j2:17]: Clippy was complaining about the documentation and formatting. The jinja template previously created:
```rust
/// Attributes for the `system` namespace.
pub mod system;
```
for example.

Then here: https://github.com/open-telemetry/weaver/blob/42203247a8217ef6909949764ab641956502e8ff/crates/weaver_codegen_test/build.rs#L84-L89 the `pub mod` lines are removed leaving a floating `///` doc comment.

File: crates/weaver_codegen_test/templates/registry/rust/metrics/mod.rs.j2 (hunk 2 above)
Comment (jerbly) [reply crates/weaver_codegen_test/templates/registry/rust/metrics/mod.rs.j2:15]: Same answer ;)

File: crates/weaver_codegen_test/templates/registry/rust/attributes/mod.rs.j2 (hunk 1 above)
Comment (lquerel) [reply crates/weaver_codegen_test/templates/registry/rust/attributes/mod.rs.j2:17]: Ah, ok. I had forgotten about that part of the plumbing in the code :-)

**Pull request merged**